### PR TITLE
AB#74821 authorization code possibility for api config

### PR DESCRIPTION
**Pull request by MwanPygmay**
# Description

- back-end support for the new api config type. Work in progress, missing the way to acquire the token
- For now, only supports adding the configuration in the database.

## Useful links

- [Please insert link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/74821/)
- [Please insert link to front-end branch if any](https://github.com/ReliefApplications/ems-frontend/pull/1999)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

see front

# Checklist:

( * == Mandatory ) 

- [x] * I have set myself as assignee of the pull request
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [x] * I have performed a self-review of my own code
- [x] * I have put the ticket for review, adding the oort-backend team to the list of reviewers
- [x] * I have commented my code, particularly in hard-to-understand areas
- [x] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [x] * I have included screenshots describing my changes if relevant
- [x] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/server/apollo/dataSources.ts [hunk 1]
```diff
@@ -180,7 +182,10 @@ export class CustomAPI extends RESTDataSource {
     if (!cacheTimestamp || cacheTimestamp < modifiedAt) {
       // Check if referenceData has changed. In this case, refresh choices instead of using cached ones.
       const body = { query: this.processQuery(referenceData) };
-      const data = await this.post(url, { body });
+      let data = await this.post(url, { body });
+      if (typeof data === 'string') {
+        data = JSON.parse(data);
+      }
```
Comment (Matthis-M-ReliefApps) [inline src/server/apollo/dataSources.ts:188]: This wasn't needed before, it is a bit annoying. I'm wondering if there hasn't been a change in RESTDataSource class with latest update that removed "didReceiveResponse" function. It seems that we were overriding it, but that it doesn't exist anymore on the base class, and I think it had the role of converting to JSON...

File: src/server/apollo/dataSources.ts [hunk 2]
```diff
@@ -21,6 +21,7 @@ const LAST_UPDATE_CODE = '{{lastUpdate}}';
 /**
  * CustomAPI class to create a dataSource fetching from an APIConfiguration.
  * If nothing is passed in the constructor, it will only be a standard REST DataSource.
+ * Datasources are called for example when using Reference Datas in a context where displaying data, for example in a grid or fetching/downloading records history items
```
Comment (Matthis-M-ReliefApps) [inline src/server/apollo/dataSources.ts:24]: Please check if this is accurate or improve description

File: src/server/apollo/dataSources.ts [hunk 3]
```diff
@@ -60,7 +61,8 @@ export class CustomAPI extends RESTDataSource {
    */
   async willSendRequest(_: string, request: AugmentedRequest) {
     if (this.apiConfiguration) {
-      const token: string = await getToken(this.apiConfiguration);
+      const accessToken = (this.server as any).req.headers.accesstoken ?? '';
+      const token: string = await getToken(this.apiConfiguration, accessToken);
```
Comment (Matthis-M-ReliefApps) [inline src/server/apollo/dataSources.ts:65]: In which scenario is willSendRequest triggered without an api configuration?

File: src/server/apollo/dataSources.ts (hunk 2 above)
Comment (AntoineRelief) [reply src/server/apollo/dataSources.ts:24]: @Matthis-M-ReliefApps 
that's a lot of "for example" but I think that's fine ;)

**Pull request merged**